### PR TITLE
Issue-122

### PR DESCRIPTION
**Pull request by Akshata-Desai-ibm**
Added logs for External Request & Response.

### CODE REVIEW

File: osvimdriver/openstack/neutron/driver.py [hunk 1]
```diff
@@ -19,14 +22,29 @@ def get_network_by_id(self, network_id):
             raise ValueError('network_id must be provided')
         neutron_client = self.__get_neutron_client()
         logger.debug('Retrieving network with id %s', network_id)
-        result = neutron_client.show_network(network_id)
-        return result['network']
+        try:
+            external_request_id = str(uuid.uuid4())
+
+            self._generate_additional_logs(network_id, 'sent', external_request_id, 'application/json',
```
Comment (lokanalla) [inline osvimdriver/openstack/neutron/driver.py:28]: network_id passed as message, is this in application/json format?

File: osvimdriver/openstack/neutron/driver.py [hunk 2]
```diff
@@ -19,14 +22,29 @@ def get_network_by_id(self, network_id):
...
+                                       'request', 'http', {'method' : 'get', 'uri' : LOG_URI_PREFIX +'/stacks'}, None)
```
Comment (lokanalla) [inline osvimdriver/openstack/neutron/driver.py:29]: is the uri verified in the library? is it the correct uri?

File: osvimdriver/openstack/neutron/driver.py [hunk 3]
```diff
@@ -19,14 +22,29 @@ def get_network_by_id(self, network_id):
...
+            raise e
+
 
     def get_network_by_name(self, network_name):
         if network_name is None:
             raise ValueError('network_name must be provided')
         neutron_client = self.__get_neutron_client()
         logger.debug('Retrieving network with name %s', network_name)
+        external_request_id = str(uuid.uuid4())
+        self._generate_additional_logs(network_name, 'sent', external_request_id, 'application/json',
```
Comment (lokanalla) [inline osvimdriver/openstack/neutron/driver.py:46]: Please check how the library sends the payload, if it in json format, the message field (natwork_name) must be put kept in Json format.

File: osvimdriver/openstack/neutron/driver.py [hunk 4]
```diff
@@ -19,14 +22,29 @@ def get_network_by_id(self, network_id):
...
+                                       'request', 'http', {'method' : 'get', 'uri' : LOG_URI_PREFIX +'/stacks'}, None)
```
Comment (lokanalla) [inline osvimdriver/openstack/neutron/driver.py:47]: Here as well the uri need to be verified with library api.

**Review by lokanalla — CHANGES_REQUESTED** [review]
Please update the message according to the payload sent by the library. 
Check the uri for each library method and update accordingly.

File: osvimdriver/openstack/heat/driver.py [hunk 5]
```diff
@@ -29,29 +34,71 @@ def create_stack(self, stack_name, heat_template, input_properties=None, files=N
...
-    def delete_stack(self, stack_id):
+        external_request_id = str(uuid.uuid4())
+
+        reqbody_dict = {"stack_name" : stack_name, "template" : heat_template, "parameters" : input_properties, "files" : files}
+        self._generate_additional_logs(reqbody_dict, 'sent', external_request_id, 'application/json',
+                                       'request', 'http', {'method' : 'post', 'uri' : LOG_URI_PREFIX +'/stacks'}, driver_request_id)
+        try:
+            create_result = heat_client.stacks.create(stack_name=stack_name, template=heat_template, parameters=input_properties, files=files)
+            stack_id = create_result['stack']['id']
+            driver_request_id = rd.build_request_id(rd.CREATE_REQUEST_PREFIX, str(stack_id))
```
Comment (lokanalla) [inline osvimdriver/openstack/heat/driver.py:46]: This driver_request_id must be returned to the caller in @ https://github.com/IBM/openstack-vim-driver/blob/develop/osvimdriver/service/resourcedriver.py. at line number 166, remove creating the driver_request_id again in line number 167, the retuned value should be passed in line number 168.

File: osvimdriver/openstack/heat/driver.py [hunk 6]
```diff
@@ -18,7 +23,7 @@ def __init__(self, session):
     def __get_heat_client(self):
         return self.__heat_client
 
-    def create_stack(self, stack_name, heat_template, input_properties=None, files=None):
+    def create_stack(self, stack_name, heat_template, input_properties=None,  files=None,driver_request_id=None ):
```
Comment (lokanalla) [inline osvimdriver/openstack/heat/driver.py:26]: create_stack function should not pass driver_request_id, as this function generates one and need to be allowed only in the response logs. request logs driver_request_id can be removed.

File: osvimdriver/openstack/neutron/driver.py [hunk 7]
```diff
@@ -19,23 +22,46 @@ def get_network_by_id(self, network_id):
             raise ValueError('network_id must be provided')
         neutron_client = self.__get_neutron_client()
         logger.debug('Retrieving network with id %s', network_id)
-        result = neutron_client.show_network(network_id)
-        return result['network']
+        try:
+            external_request_id = str(uuid.uuid4())
+            driver_request_id  = str(uuid.uuid4())
```
Comment (lokanalla) [inline osvimdriver/openstack/neutron/driver.py:27]: driver_request_id is specific to a lifecycle, Please pass this value from the calling function.

File: osvimdriver/openstack/neutron/driver.py [hunk 8]
```diff
@@ -19,23 +22,46 @@ def get_network_by_id(self, network_id):
...
+
 
     def get_network_by_name(self, network_name):
         if network_name is None:
             raise ValueError('network_name must be provided')
         neutron_client = self.__get_neutron_client()
         logger.debug('Retrieving network with name %s', network_name)
-        result = neutron_client.list_networks()
+        external_request_id = str(uuid.uuid4())
+        driver_request_id  = str(uuid.uuid4())
```
Comment (lokanalla) [inline osvimdriver/openstack/neutron/driver.py:46]: driver_request_id is specific to a lifecycle, Please pass this value from the calling function.

File: osvimdriver/openstack/neutron/driver.py [hunk 9]
```diff
@@ -45,5 +71,53 @@ def get_subnet_by_id(self, subnet_id):
             raise ValueError('subnet_id must be provided')
         neutron_client = self.__get_neutron_client()
         logger.debug('Retrieving subnet with id %s', subnet_id)
-        result = neutron_client.show_subnet(subnet_id)
-        return result['subnet']
+        try:
+            external_request_id = str(uuid.uuid4())
+            driver_request_id  = str(uuid.uuid4())
```
Comment (lokanalla) [inline osvimdriver/openstack/neutron/driver.py:76]: driver_request_id is specific to a lifecycle, Please pass this value from the calling function.

**Review by lokanalla — CHANGES_REQUESTED** [review]
I see that the function _generate_additional_logs()defined in 2 files, can't you create a separate module and use it both the files.

File: osvimdriver/openstack/neutron/driver.py (hunk 7 above)
Comment (Akshata-Desai-ibm) [reply osvimdriver/openstack/neutron/driver.py]: I have done changes.

File: osvimdriver/openstack/neutron/driver.py (hunk 8 above)
Comment (Akshata-Desai-ibm) [reply osvimdriver/openstack/neutron/driver.py]: I have done changes.

File: osvimdriver/openstack/neutron/driver.py (hunk 9 above)
Comment (Akshata-Desai-ibm) [reply osvimdriver/openstack/neutron/driver.py]: I have done changes.

File: osvimdriver/openstack/heat/driver.py (hunk 6 above)
Comment (Akshata-Desai-ibm) [reply osvimdriver/openstack/heat/driver.py]: I have done changes.

File: osvimdriver/openstack/heat/driver.py [hunk 10]
```diff
@@ -29,29 +35,71 @@ def create_stack(self, stack_name, heat_template, input_properties=None, files=N
...
+            common._generate_additional_logs(create_result, 'received', external_request_id, 'application/json',
+                                       'response', 'http', {'status_code' : 201,'status_reason_phrase' : 'Created'}, driver_request_id)
+            logger.debug('Stack with name %s created and assigned id %s', stack_name, stack_id)
+            return stack_id
+        except (heatexc.HTTPNotFound,heatexc.HTTPBadRequest) as e:
+            status_reason_phrase = 'Not Found'
+            if  e.code != 404:
+                status_reason_phrase = 'Bad Request'
+            common._generate_additional_logs(e, 'received', external_request_id, 'plain/text',
+                                       'response', 'http', {'status_code' : e.code,'status_reason_phrase' : status_reason_phrase}, driver_request_id)
```
[the remark this answers is not in the document]
Comment (Akshata-Desai-ibm) [reply osvimdriver/openstack/heat/driver.py:61]: I have done changes.

File: osvimdriver/openstack/heat/driver.py (hunk 5 above)
Comment (Akshata-Desai-ibm) [reply osvimdriver/openstack/heat/driver.py:49]: I have done changes.

File: osvimdriver/openstack/heat/driver.py [hunk 11]
```diff
@@ -29,29 +35,75 @@ def create_stack(self, stack_name, heat_template, input_properties=None, files=N
...
+            external_request_id = str(uuid.uuid4())
+           
+            common._generate_additional_logs('', 'sent', external_request_id, '',
+                                        'request', 'http', {'method':'delete', 'uri' : LOG_URI_PREFIX + '/stacks/' + stack_id}, driver_request_id)
             delete_result = heat_client.stacks.delete(stack_id)
-        except heatexc.HTTPNotFound as e:
+            result = ''
+            if delete_result != None:   
+                result = delete_result
+            common._generate_additional_logs(result, 'received', external_request_id, '',
```
Comment (lokanalla) [inline osvimdriver/openstack/heat/driver.py:78]: Would the content type be '', if the delete_result != None. You may need to handle the other case here for content type.

File: osvimdriver/openstack/neutron/driver.py [hunk 12]
```diff
@@ -14,36 +18,69 @@ def __init__(self, session):
...
         neutron_client = self.__get_neutron_client()
         logger.debug('Retrieving network with id %s', network_id)
-        result = neutron_client.show_network(network_id)
-        return result['network']
+        try:
+            external_request_id = str(uuid.uuid4())
+            common._generate_additional_logs('', 'sent', external_request_id, '',
+                                       'request', 'http', {'method' : 'get', 'uri' : LOG_URI_PREFIX +'/networks/' + network_id }, driver_request_id)
+            result = neutron_client.show_network(network_id)
+            common._generate_additional_logs(result, 'received', external_request_id, 'plain/text',
```
Comment (lokanalla) [inline osvimdriver/openstack/neutron/driver.py:31]: I am expecting the library would return the values in dict/json. May be you can update the content type to application/json and convert the result to str.

File: osvimdriver/openstack/neutron/driver.py [hunk 13]
```diff
@@ -14,36 +18,69 @@ def __init__(self, session):
...
             raise ValueError('network_name must be provided')
         neutron_client = self.__get_neutron_client()
         logger.debug('Retrieving network with name %s', network_name)
-        result = neutron_client.list_networks()
+        external_request_id = str(uuid.uuid4())
+        common._generate_additional_logs('', 'sent', external_request_id, '',
+                                       'request', 'http', {'method' : 'get', 'uri' : LOG_URI_PREFIX +'/networks' }, driver_request_id)
+        try:
+            result = neutron_client.list_networks()
+            common._generate_additional_logs(str(result), 'received', external_request_id, 'plain/text',
```
Comment (lokanalla) [inline osvimdriver/openstack/neutron/driver.py:50]: I am expecting the library would return the values in dict/json. May be you can update the content type to application/json and here you have already converted the result to str.

File: osvimdriver/openstack/neutron/driver.py [hunk 14]
```diff
@@ -14,36 +18,69 @@ def __init__(self, session):
...
         neutron_client = self.__get_neutron_client()
         logger.debug('Retrieving subnet with id %s', subnet_id)
-        result = neutron_client.show_subnet(subnet_id)
-        return result['subnet']
+        try:
+            external_request_id = str(uuid.uuid4())
+            common._generate_additional_logs('', 'sent', external_request_id, '',
+                                       'request', 'http', {'method' : 'get', 'uri' : LOG_URI_PREFIX +'/subnets/' + subnet_id}, driver_request_id)
+            result = neutron_client.show_subnet(subnet_id)
+            common._generate_additional_logs(result, 'received', external_request_id, 'plain/text',
```
Comment (lokanalla) [inline osvimdriver/openstack/neutron/driver.py:78]: I am expecting the library would return the values in dict/json. May be you can update the content type to application/json and convert the result to str.

**Review by lokanalla — CHANGES_REQUESTED** [review]
Please do the suggested changes.

File: osvimdriver/openstack/heat/driver.py (hunk 11 above)
Comment (Akshata-Desai-ibm) [reply osvimdriver/openstack/heat/driver.py]: I have made changes.

File: osvimdriver/openstack/neutron/driver.py (hunk 12 above)
Comment (Akshata-Desai-ibm) [reply osvimdriver/openstack/neutron/driver.py]: I have made changes.

File: osvimdriver/openstack/neutron/driver.py (hunk 13 above)
Comment (Akshata-Desai-ibm) [reply osvimdriver/openstack/neutron/driver.py]: I have made changes.

File: osvimdriver/openstack/neutron/driver.py (hunk 14 above)
Comment (Akshata-Desai-ibm) [reply osvimdriver/openstack/neutron/driver.py]: I have made changes.

**Review by lokanalla — CHANGES_REQUESTED** [review]
In the file https://github.com/Akshata-Desai-ibm/openstack-vim-driver/blob/issue/122/osvimdriver/openstack/heat/driver.py,

I see that for create and delete stack results are not converted to str but get result is converted to str.

Please make sure consistent across APIs and also see the result logs are properly coming.

**Review by lokanalla — APPROVED** [review]

**Review by lokanalla — APPROVED** [review]

**Pull request merged**